src/Storages/StorageBuffer.cpp

```diff
@@ -841,7 +841,7 @@ void StorageBuffer::flushBuffer(Buffer & buffer, bool check_thresholds, bool loc
 
     size_t block_rows = block_to_write.rows();
     size_t block_bytes = block_to_write.bytes();
-    size_t block_allocated_bytes = block_to_write.allocatedBytes();
+    size_t block_allocated_bytes_delta = block_to_write.allocatedBytes() - buffer.data.allocatedBytes();
```

Comment:
This does not looks correct, since `buffer.data` is empty here, see two lines above:

```c
    buffer.data.swap(block_to_write);
```

Seems that the problem is in different place?
I left few comments for the original PR #24066

Do you have a reproducer?
It looks like this case can be covered with a test.

Reply:
Buffer.data is empty (has no rows), but according to the code, allocatedBytes() can be non-zero. For example, [allocatedBytes](https://github.com/ClickHouse/ClickHouse/blob/9f5cd35a6963cc556a51218b46b0754dcac7306a/src/Columns/ColumnFixedString.h#L65) for FixedString returns at least sizeof(size_t). And [allocated_bytes](https://github.com/ClickHouse/ClickHouse/blob/efe70aaa076002e33fc79d24a1d3719b658ac3ea/src/Common/PODArray.h#L215) in PODArray returns at least pad_right + pad_left bytes.

Reply:
>Buffer.data is empty (has no rows), 

`buffer.data` is not Column, but block, and it's `swap` method is:

https://github.com/ClickHouse/ClickHouse/blob/9f5cd35a6963cc556a51218b46b0754dcac7306a/src/Core/Block.cpp#L623-L628

So it just moves the `data`, and so `Block::allocatedBytes` will return 0 after this:

https://github.com/ClickHouse/ClickHouse/blob/9f5cd35a6963cc556a51218b46b0754dcac7306a/src/Core/Block.cpp#L264-L271

Reply:
block_to_write was created as cloneEmpty().
https://github.com/ClickHouse/ClickHouse/blob/d4998909b666e7129ca14221e6546b457086e16f/src/Core/Block.cpp#L313-L321
Thus, block_to_write.data is not empty. Instead, every column in this block is empty. And for these columns allocatedBytes() can be non-zero.

Reply:
Oh, I see, you are right, thanks!
BTW seems that this can be covered with a test, can you take a look?

Reply:
Yes. I'll add it today